src/query/shorthand.ts

```diff
@@ -255,3 +258,184 @@ export function fieldDef(encQ: EncodingQuery,
   fieldAndParams += props.map((p) => ',' + p.key + '=' + p.value).join('');
   return (fn ? fn + '(' + fieldAndParams + ')' : fieldAndParams);
 }
+
+const CHANNEL_INDEX = toMap(CHANNELS);
+const AGGREGATE_OP_INDEX = toMap(AGGREGATE_OPS);
+const SINGLE_TIMEUNIT_INDEX = toMap(SINGLE_TIMEUNITS);
+const MULTI_TIMEUNIT_INDEX = toMap(MULTI_TIMEUNITS);
+
+export function parse(shorthand: string): SpecQuery {
+  let splitShorthand = shorthand.split('|');
+
+  let specQ: SpecQuery = {mark: undefined, encodings: [] as EncodingQuery[]};
+
+  // Mark
+  specQ.mark = splitShorthand[0];
+
+  for (let part of splitShorthand) {
+    const splitPart = splitWithTail(part, ':', 1);
+    const splitPartKey = splitPart[0];
+    const splitPartValue = splitPart[1];
+
+    if (CHANNEL_INDEX[splitPartKey]) {
+      shorthandParser.encoding(specQ, splitPartKey, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'calculate') {
+      shorthandParser.calculate(specQ, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'filter') {
+      shorthandParser.filter(specQ, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'filterInvalid') {
+      shorthandParser.filterInvalid(specQ, splitPartValue);
+      continue;
+    }
+  }
+
+  return specQ;
+}
+
+/**
+ * Split a string n times into substrings with the specified delimiter and return them as an array.
+ * @param str The string to be split
+ * @param delim The delimiter string used to separate the string
+ * @param number The value used to determine how many times the string is split
+ */
+export function splitWithTail(str: string, delim: string, count: number) {
+  let parts = str.split(delim);
+  let tail = parts.slice(count).join(delim);
+  let result = parts.slice(0, count);
+  result.push(tail);
+  return result;
+}
+
+export namespace shorthandParser {
+  export function encoding(specQ: SpecQuery, channel: string, fieldDefShorthand: string) {
+    let encQ: EncodingQuery = {channel: channel};
+
+    if (fieldDefShorthand.indexOf('(') !== -1) {
+      encQ = fn(encQ, fieldDefShorthand);
+    } else {
+      encQ = rawFieldDef(encQ, splitWithTail(fieldDefShorthand, ',', 2));
+    }
+
+    specQ.encodings.push(encQ);
+  }
+
+  export function rawFieldDef(encQ: EncodingQuery, fieldDefPart: string[]): EncodingQuery {
+    encQ.field = fieldDefPart[0];
+    encQ.type = TYPE_FROM_SHORT_TYPE[fieldDefPart[1].toUpperCase()];
+
+    let partParams = fieldDefPart[2];
+    let closingBraceIndex = 0;
+
+    // TODO: don't use a loop
```

Comment:
Change TODO to FIXME.   (yes, this is not very good and i'd not let this slip if it's not a deadline.)